rusk-wallet/src/imp.rs

```diff
@@ -0,0 +1,299 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+//
+// Copyright (c) DUSK NETWORK. All rights reserved.
+
+use crate::tx::UnprovenTransaction;
+use crate::{NodeClient, Store, Transaction};
+
+use alloc::vec::Vec;
+
+use canonical::CanonError;
+use dusk_bytes::Error as BytesError;
+use dusk_jubjub::{BlsScalar, JubJubScalar};
+use dusk_pki::PublicSpendKey;
+use phoenix_core::{Crossover, Error as PhoenixError, Fee, Note, NoteType};
+use rand_core::{CryptoRng, Error as RngError, RngCore};
+
+const MAX_INPUT_NOTES: usize = 0x4;
+
+/// The error type returned by this crate.
+#[derive(Debug)]
+pub enum Error<S: Store, C: NodeClient> {
+    /// Underlying store error.
+    Store(S::Error),
+    /// Error originating from the node client.
+    Node(C::Error),
+    /// Canonical stores.
+    Canon(CanonError),
+    /// Random number generator error.
+    Rng(RngError),
+    /// Serialization and deserialization of Dusk types.
+    Bytes(BytesError),
+    /// Originating from the transaction model.
+    Phoenix(PhoenixError),
+    /// Not enough balance to perform transaction.
+    NotEnoughBalance,
+    /// Note combination for the given value is impossible given the maximum
+    /// amount if inputs in a transaction.
+    NoteCombinationProblem,
+}
+
+impl<S: Store, C: NodeClient> Error<S, C> {
+    /// Returns an error from the underlying store error.
+    pub fn from_store_err(se: S::Error) -> Self {
+        Self::Store(se)
+    }
+    /// Returns an error from the underlying note finder error.
+    pub fn from_node_err(ne: C::Error) -> Self {
+        Self::Node(ne)
+    }
+}
```

Comment:
Yeah, if I try to implement both:
 
```rust
impl<C: NodeClient, S: Store> From<S::Error> for Error<S, C> {
    fn from(se: S::Error) -> Self {
        Self::Store(se)
    }
}
impl<C: NodeClient, S: Store> From<C::Error> for Error<S, C> {
    fn from(ne: C::Error) -> Self {
        Self::Node(ne)
    }
}
```

I get into trouble because both `C::Error` and `S::Error` can conflict with each other and other `From` implementations.